### PR TITLE
Feat/save to mydrive

### PR DESCRIPTION
**Pull request by maya-fisher**

### CODE REVIEW

File: src/api/files.js [hunk 1]
```diff
@@ -261,3 +262,40 @@ export function getPreview(fileID) {
...
+export async function GetCopiedFileName(fileID, folderID) {
+  const file = await getFileByID(fileID);
+    if (folderID == undefined) {
+      const folderFiles = await fetchFiles();
+      return await getNewFileName(file.name, folderFiles);
+    } else {
+      const folder = await getFileByID(folderID);
+      const folderFiles = await fetchFiles(folder);
+      return await getNewFileName(file.name, folderFiles);
+    }
```
Comment (yoavrozner) [inline src/api/files.js:282]: Is this the real life? Is this just fantasy?

File: src/api/files.js [hunk 2]
```diff
@@ -261,3 +262,40 @@ export function getPreview(fileID) {
...
+    files: folders,
+    loadingFiles: [],
+  });
+
+  
+  if (isExist) {
+    return newFileName;
+  } else {
+    return name;
+  }
```
Comment (yoavrozner) [inline src/api/files.js:300]:
```suggestion
  const [isExist, newFileName] = appendNumberIfFileExists({
    name: name,
    files: folders,
    loadingFiles: [],
  });

  return isExist ? newFileName : name; 
```

File: src/api/files.js [hunk 3]
```diff
@@ -261,3 +262,40 @@ export function getPreview(fileID) {
...
+    loadingFiles: [],
+  });
+
+  
+  if (isExist) {
+    return newFileName;
+  } else {
+    return name;
+  } 
+}
```
Comment (yoavrozner) [inline src/api/files.js:301]: Missing newline

File: src/components/buttons/CopyFileButton.vue [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    <CopyFilePopup ref="copyPopup" :files="chosenFiles" @confirm="onSubmit" />
+    <span>{{ $t("buttons.MakeCopy") }}</span>
+  </v-tooltip>
+</template>
+
+<script>
+import { mapGetters } from "vuex";
+import { writeRole } from "@/utils/roles";
+import CopyFilePopup from "@/components/popups/CopyFilePopup";
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:26]: Default imports comes first!

File: src/components/buttons/CopyFileButton.vue [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  name: "CopyFileButton",
+  props: ["icon"],
+  components: { CopyFilePopup },
+  computed: {
+    ...mapGetters(["chosenFiles", "currentFolder"]),
+  },
+  data() {
+    return {
+    };
+  },
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:39]: Redundant

File: src/components/buttons/CopyFileButton.vue [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  methods: {
+    copyButtonAppear() {
+      return this.chosenFiles.every((file) => !isFolder(file.type)) ? true : false;
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:42]:
```suggestion
      return this.chosenFiles.every((file) => !isFolder(file.type));
```

Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:42]: Like every! well done!

Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:42]: && is single file

File: src/components/buttons/CopyFileButton.vue [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        folderID,
+      });
+
+    // folderID ? this.$router.push({ path: "/folders", query: { id: folderID } }) 
+    // : this.$router.currentRoute.path === "/favorites" ? this.$router.push("/my-drive") :
+
+      if (folderID) {
+        this.$router.push({ path: "/folders", query: { id: folderID } });
+      } else if (this.$router.currentRoute.path === "/favorites" || this.$router.currentRoute.path === "/shared-with-me") {
+          this.$router.push("/my-drive")
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:56]: So help me G*d

File: src/components/buttons/CopyFileButton.vue [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      }
+      this.$emit("close");
+    },
+    canMove() {
+      return (
+        (!(this.currentFolder || this.isSharedFile()) ||
+          (this.currentFolder && writeRole(this.currentFolder.role))) &&
+        this.chosenFiles.every((file) => writeRole(file.role)) &&
+        !this.isFileReadOnly()
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:65]: Put it in the utils

File: src/components/buttons/CopyFileButton.vue [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      return firstFile && firstFile.shared;
+    },
+    isFileReadOnly() {
+      return this.chosenFiles.every(
+        (file) => file?.isReadOnly != undefined && file.isReadOnly
+      );
+    },
+  },
+};
+</script>
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:82]: New line

File: src/components/popups/CopyFilePopup.vue [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+    async open(buttonIndex) {
+      this.copyButton = buttonIndex;
+      await Promise.all([this.fetchHierachy(this.currentFolder), this.fetchFolders(this.currentFolder)]);
+      this.dialog = true;
+    },
+    async opensharedFolders(buttonIndex) {
+      this.copyButton = buttonIndex;
+      await Promise.all([this.fetchHierachy(this.currentFolder), this.fetchSharedFolders(this.currentFolder)]);
+      this.dialog = true;
+    },
```
Comment (yoavrozner) [inline src/components/popups/CopyFilePopup.vue:72]: One function to rule them all

File: src/components/popups/CopyFilePopup.vue [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    async fetchFolders(parent) {
+        this.currentChildren = await filesApi.getFoldersByFolder(parent ? parent.id : undefined);
+    },
+    async fetchSharedFolders(parent) {
+      if (parent === undefined) {
+        this.submitButtonDisabled = true;
+        this.currentChildren = await filesApi.fetchSharedFolders(parent ? parent.id : undefined);
+      } else {
+        this.submitButtonDisabled = false;
+        this.currentChildren = await filesApi.getFoldersByFolder(parent ? parent.id : undefined);
```
Comment (yoavrozner) [inline src/components/popups/CopyFilePopup.vue:82]: ?

File: src/components/popups/CopyFilePopup.vue [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+            value: undefined,
+            text: this.$t("pageHeaders.MyDrive"),
+            disabled: !folder,
+          });
+        } else {
+          breadcrumbs.push({
+            value: undefined,
+            text: this.$t("pageHeaders.SharedWithMe"),
+            disabled: !folder,
+          });
```
Comment (yoavrozner) [inline src/components/popups/CopyFilePopup.vue:105]: Switch case with enum

File: src/components/popups/CopyFilePopup.vue [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+      }
+
+      this.folderHierarchy = breadcrumbs;
+    },
+    async onFolderChange(folder) {
+      if (this.isFolderInFolder(folder)) return;
+      await this.fetchHierachy(folder);
+      if (this.copyButton === 0) await this.fetchFolders(folder);
+      else await this.fetchSharedFolders(folder);
+      this.folderDest = folder;
```
Comment (yoavrozner) [inline src/components/popups/CopyFilePopup.vue:138]: Switch case

File: src/store/modules/files.js [hunk 14]
```diff
@@ -58,31 +58,31 @@ const actions = {
...
 
-    /**
-   * addFav gets a file id and add it to favorites
-   * @param fileID is the id of the file to add
-   */
+/**
+* addFav gets a file id and add it to favorites
+* @param fileID is the id of the file to add
+*/
   async addFav({dispatch }, file) {
```
Comment (yoavrozner) [inline src/store/modules/files.js:65]:
```suggestion
  async addFav({ dispatch }, file) {
```

**Review by yoavrozner — CHANGES_REQUESTED** [review]
Well done @maya-fisher !
Please see my notes

File: src/api/files.js [hunk 15]
```diff
@@ -3,6 +3,7 @@ import { pushUpdatedFile, removeUpdatedFile } from "@/utils/lastUpdatedFileHandl
 import { baseURL, fileTypes, pageSize } from "@/config";
 import { isFolder } from "@/utils/isFolder";
 import store from "@/store";
```
[the remark this answers is not in the document]
Comment (maya-fisher) [reply src/api/files.js:5]: Done

File: src/api/files.js [hunk 16]
```diff
@@ -261,3 +262,40 @@ export function getPreview(fileID) {
 export function getPdfPreview(fileID) {
   return `${baseURL}/api/files/${fileID}?alt=media&preview`;
 }
+
+export async function copyFile(fileID, folderID, newFileName) {
+  const res = await Axios.post(`${baseURL}/api/files/copyObject/${fileID}/${folderID}/${newFileName}?appId=drive`)
+  return res;
```
[the remark this answers is not in the document]
Comment (maya-fisher) [reply src/api/files.js:268]: Done

File: src/api/files.js (hunk 1 above)
Comment (maya-fisher) [reply src/api/files.js:282]: fantasy

File: src/api/files.js (hunk 2 above)
Comment (maya-fisher) [reply src/api/files.js:300]: Done

File: src/api/files.js (hunk 3 above)
Comment (maya-fisher) [reply src/api/files.js:301]: Done

File: src/components/buttons/CopyFileButton.vue (hunk 4 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:26]: Done

File: src/components/buttons/CopyFileButton.vue (hunk 5 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:39]: Done

File: src/components/buttons/CopyFileButton.vue (hunk 6 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:42]: Done

Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:42]: Thanks! I think I copied this from favorites

File: src/components/buttons/CopyFileButton.vue (hunk 7 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:56]: :(

File: src/components/buttons/CopyFileButton.vue (hunk 8 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:65]: I forgot it's not necessary here at all

File: src/components/buttons/CopyFileButton.vue (hunk 9 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:82]: Done

File: src/components/popups/CopyFilePopup.vue (hunk 10 above)
Comment (maya-fisher) [reply src/components/popups/CopyFilePopup.vue:72]: Done

File: src/components/popups/CopyFilePopup.vue (hunk 12 above)
Comment (maya-fisher) [reply src/components/popups/CopyFilePopup.vue:105]: Done

File: src/components/popups/CopyFilePopup.vue (hunk 13 above)
Comment (maya-fisher) [reply src/components/popups/CopyFilePopup.vue:138]: Done

File: src/store/modules/files.js (hunk 14 above)
Comment (maya-fisher) [reply src/store/modules/files.js:65]: Done

File: src/components/popups/CopyFilePopup.vue (hunk 11 above)
Comment (maya-fisher) [reply src/components/popups/CopyFilePopup.vue:82]: To copy a file into shared with me, you have to select a folder.

Comment (maya-fisher) [reply src/components/popups/CopyFilePopup.vue:82]: If there is something else that is wrong then what it is?

File: src/api/files.js [hunk 17]
```diff
@@ -261,3 +262,35 @@ export function getPreview(fileID) {
...
+export async function GetCopiedFileName(fileID, folderID) {
+  const file = await getFileByID(fileID);
+    if (folderID == undefined) {
+      const folderFiles = await fetchFiles();
+      return await getNewFileName(file.name, folderFiles);
+    } else {
+      const folder = await getFileByID(folderID);
+      const folderFiles = await fetchFiles(folder);
+      return await getNewFileName(file.name, folderFiles);
+    }
```
Comment (yoavrozner) [inline src/api/files.js:282]:
```suggestion
    if (folderID == undefined) {
      const [file, folderFiles] = await Promise.all([getFileByID(fileID), fetchFiles()]);
      return await getNewFileName(file.name, folderFiles);
    } else {
      const folder = await getFileByID(folderID);
      const [file, folderFiles] = await Promise.all([getFileByID(fileID), fetchFiles(folder)]);
      return await getNewFileName(file.name, folderFiles);
    }
```

File: src/api/files.js [hunk 18]
```diff
@@ -261,3 +262,35 @@ export function getPreview(fileID) {
...
+}
+
+// The function gets the file name and the files in a folder and checks if the file name already exists.
+// If it does, it returns the new file name with a number appended to it.
+// If it doesn't, it returns the original file name.
+async function getNewFileName(name, folders) {
+  let [isExist, newFileName] = appendNumberIfFileExists({
```
Comment (yoavrozner) [inline src/api/files.js:289]:
```suggestion
  const [isExist, newFileName] = appendNumberIfFileExists({
```

File: src/components/buttons/CopyFileButton.vue [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+      this.$store.dispatch("copyFile", {
+        fileID: this.chosenFiles[0].id,
+        folderID,
+      });
+
+      if (folderID) {
+        this.$router.push({ path: "/folders", query: { id: folderID } });
+      } else if (this.$router.currentRoute.path === "/favorites" || this.$router.currentRoute.path === "/shared-with-me") {
+          this.$router.push("/my-drive")
+      }
```
Comment (yoavrozner) [inline src/components/buttons/CopyFileButton.vue:49]: The indentations are not equal

**Review by yoavrozner — COMMENTED** [review]
Well done @maya-fisher !
Please see my notes

File: src/api/files.js (hunk 17 above)
Comment (maya-fisher) [reply src/api/files.js:282]: Ok.

File: src/api/files.js (hunk 18 above)
Comment (maya-fisher) [reply src/api/files.js:289]: Done.

File: src/components/buttons/CopyFileButton.vue (hunk 19 above)
Comment (maya-fisher) [reply src/components/buttons/CopyFileButton.vue:49]: I didn't understand what you meant.

File: docker-compose.yml [hunk 20]
```diff
@@ -8,7 +8,7 @@ services:
     depends_on:
       - api-gateway
   api-gateway:
-    image: "drivehub.azurecr.io/meateam/api-gateway:develop"
+    image: "drivehub.azurecr.io/meateam/api-gateway:save-to-mydrive"
```
[the remark this answers is not in the document]
Comment (maya-fisher) [reply docker-compose.yml]: Done. But need to push to drivehub the updated image

**Review by yoavrozner — APPROVED** [review]
💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯 💯

File: src/components/buttons/CopyFileButton.vue (hunk 19 above)
Comment (yoavrozner) [reply src/components/buttons/CopyFileButton.vue:49]: בעברית, ההזחה שלך לא שווה בין התנאי והתנאי השני

File: docker-compose.yml (hunk 20 above)
[the remark this answers is not in the document]
Comment (yoavrozner) [reply docker-compose.yml]: In the ui we have ci that auto pushes the new image to the drivehub

**Review by yarinBenisty — APPROVED** [review]

**Pull request merged**